metronome/networking/src/io/iohk/metronome/networking/RemoteConnectionManager.scala

```diff
@@ -0,0 +1,339 @@
+          connection
+            .close()
+            .flatMap(_ =>
+              Sync[F].raiseError(ConnectionAlreadyClosedException(recipient))
+            )
+        }
+      case None =>
+        Sync[F].raiseError(ConnectionAlreadyClosedException(recipient))
+    }
+  }
+}
+//TODO add logging
+object RemoteConnectionManager {
+  case class ConnectionAlreadyClosedException[K](key: K)
+      extends RuntimeException(
+        s"Connection with node ${key}, has already closed"
+      )
+
+  private def getConnectionErrorMessage[K](
+      e: ConnectionError,
+      connectionKey: K
+  ): String = {
+    e match {
+      case EncryptedConnectionProvider.DecodingError =>
+        s"Unexpected decoding error on connection with ${connectionKey}"
+      case EncryptedConnectionProvider.UnexpectedError(ex) =>
+        s"Unexpected error ${ex.getMessage} on connection with ${connectionKey}"
+    }
+  }
+
+  case class UnexpectedConnectionError[K](e: ConnectionError, connectionKey: K)
+      extends RuntimeException(getConnectionErrorMessage(e, connectionKey))
+
+  case class MessageReceived[K, M](from: K, message: M)
+
+  case class ConnectionSuccess[F[_], K, M](
+      encryptedConnection: EncryptedConnection[F, K, M]
+  )
+
+  case class ConnectionFailure[K](
+      connectionRequest: OutGoingConnectionRequest[K],
+      err: Throwable
+  )
+
+  private def connectTo[
+      F[_]: Sync,
+      K: Codec,
+      M: Codec
+  ](
+      encryptedConnectionProvider: EncryptedConnectionProvider[F, K, M],
+      connectionRequest: OutGoingConnectionRequest[K]
+  ): F[Either[ConnectionFailure[K], ConnectionSuccess[F, K, M]]] = {
+    encryptedConnectionProvider
+      .connectTo(connectionRequest.key, connectionRequest.address)
+      .redeemWith(
+        e => Sync[F].pure(Left(ConnectionFailure(connectionRequest, e))),
+        connection => Sync[F].pure(Right(ConnectionSuccess(connection)))
+      )
+  }
+
+  case class RetryConfig(
+      initialDelay: FiniteDuration,
+      backOffFactor: Long,
+      maxDelay: FiniteDuration,
+      maxRandomJitter: Option[FiniteDuration]
+  )
+
+  object RetryConfig {
+    import scala.concurrent.duration._
+    def default: RetryConfig = {
+      RetryConfig(500.milliseconds, 2, 30.seconds, maxRandomJitter = None)
+    }
+
+  }
+
+  private def retryConnection[F[_]: Timer: Concurrent, K](
+      config: RetryConfig,
+      failedConnectionRequest: OutGoingConnectionRequest[K]
+  ): F[OutGoingConnectionRequest[K]] = {
+    val updatedFailureCount =
+      failedConnectionRequest.numberOfFailures + 1
+    val exponentialBackoff =
+      math.pow(config.backOffFactor.toDouble, updatedFailureCount).toLong
+
+    val randomJitter =
+      config.maxRandomJitter.fold(FiniteDuration(0, TimeUnit.MILLISECONDS)) {
+        maxJitter =>
+          val randomJitterInMillis =
+            ThreadLocalRandom.current().nextLong(maxJitter.toMillis)
+          FiniteDuration(randomJitterInMillis, TimeUnit.MILLISECONDS)
```

Comment:
Maybe it could be a percentage value, to scale with the delay itself, e.g. +/-10%, so 500ms would be 450-550ms, at 30s it would be 27-33s.

Reply:
neat idea ! :+1: